components/camel-reactive-streams/src/main/java/org/apache/camel/component/reactive/streams/api/CamelReactiveStreams.java

```diff
@@ -0,0 +1,166 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.reactive.streams.api;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.camel.CamelContext;
+import org.apache.camel.Exchange;
+import org.apache.camel.component.reactive.streams.engine.CamelReactiveStreamsServiceImpl;
+import org.apache.camel.spi.FactoryFinder;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is the main entry-point for getting Camel streams associate to reactive-streams endpoints.
+ *
+ * It delegates main methods to an instance of {@link CamelReactiveStreamsService}. This component provides
+ * a default implementation that can be overridden in a 'META-INF/services/reactive-streams/reactiveStreamsService' file.
+ */
+public final class CamelReactiveStreams {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CamelReactiveStreams.class);
+
+    private static Map<CamelContext, CamelReactiveStreams> instances = new HashMap<>();
+
+    private CamelReactiveStreamsService service;
+
+    private CamelReactiveStreams(CamelReactiveStreamsService service) {
+        this.service = service;
+    }
+
+    public static synchronized CamelReactiveStreams get(CamelContext context) {
+        if (!instances.containsKey(context)) {
+            CamelReactiveStreamsService service = resolveReactiveStreamsService(context);
+            try {
+                context.addService(service, true, true);
+            } catch (Exception ex) {
+                throw new IllegalStateException("Cannot add the CamelReactiveStreamsService to the Camel context", ex);
+            }
+
+            instances.put(context, new CamelReactiveStreams(service));
+        }
+
+        return instances.get(context);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static CamelReactiveStreamsService resolveReactiveStreamsService(CamelContext context) {
+        Class<? extends CamelReactiveStreamsService> serviceClass = null;
+        try {
+            FactoryFinder finder = context.getFactoryFinder("META-INF/services/reactive-streams/");
+            LOG.trace("Using FactoryFinder: {}", finder);
+            serviceClass = (Class<? extends CamelReactiveStreamsService>) finder.findClass("reactiveStreamsService");
+        } catch (ClassNotFoundException e) {
```

Comment:
There is a small issue with this factory finder which is that you may have multiple component providing an implementation of CamelReactiveStreamsService and with the current implementation you'll end up using the first one found in the classpath, which may not be what you'd expect.

It would be nice to have a way to:

- lookup the factory by name i.e CamelReactiveStreams.get(context, "rx-java"), something like:

```java
    // Lookup in context registry
    CamelReactiveStreamsService crss = context.getRegistry().lookupByNameAndType(name, CamelReactiveStreamsService.class);

    if (crss == null) {
        // lookup with provided name, if not provided fallback to default service name
        crss = resolveReactiveStreamsService(context, new String[] { name, "reactive-stream-service" })
    }
```

- have a way to set the type in the endpoint uri

```java
    from("reactive-streams:id?#serviceType=rx-java")
        .to("...")
```